File: Strings/leetcode14.py
from typing import List


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ""

        prefix = strs[0]

        for i in range(1, len(strs)):
            s = strs[i]
            j = 0
            while j < len(prefix) and j < len(s) and prefix[j] == s[j]:
                j += 1
            prefix = prefix[:j]
            if prefix == "":
                break

        return prefix


if __name__ == "__main__":
    solution = Solution()
    strs = ["flower", "flow", "flight"]
    print(solution.longestCommonPrefix(strs))
    strs = ["dog", "racecar", "car"]
    print(solution.longestCommonPrefix(strs))

File: Strings/leetcode14.cpp
// Write a function to find the longest common prefix string amongst an array of strings.
// If there is no common prefix, return an empty string "".
// Example 1:
// Input: strs = ["flower","flow","flight"]
// Output: "fl"
// Example 2:
// Input: strs = ["dog","racecar","car"]
// Output: ""
// Explanation: There is no common prefix among the input strings.

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        string ans="";
        sort(strs.begin(),strs.end());
        int n=strs.size();
        string first=strs[0],last=strs[n-1];
        for(int i=0;i<min(first.size(),last.size());i++){
            if(first[i]!=last[i]){
                return ans;
            }
            ans+=last[i];
        }
        return ans;
    }
};

int main(){
    Solution s;
    vector<string> strs={"flowers","flow","flo"};
    cout << s.longestCommonPrefix(strs);
    return 0;
}